### PR TITLE
Replace PointerArray<T> with type alias to std::vector<T*>

### PR DESCRIPTION
**Pull request by rprospero**
This completely removes the PointerArray class and replaces it with std::vector.  Since the method names were not identical between the two classes, this required some in place editing where PointerArray had been used.

As port of the modernization, all the locations where the nItems() member function was used to set the limits on a for loop has been replaced with a proper range based for loop.

This PR is based on the feedback that @LamarMoore provided on pr #217.  At most, only one of these two requests should be merged.

### CODE REVIEW

File: src/gui/viewer_render.cpp [hunk 1]
```diff
@@ -371,9 +371,8 @@ void BaseViewer::postRedisplay()
 	PointerArray<BaseViewer> updateStack;
 	createUpdateStack(updateStack);
 
-	for (int n=0; n<updateStack.nItems(); ++n)
+	for (auto viewer : updateStack)
```
Comment (LamarMoore) [inline src/gui/viewer_render.cpp]: For the range based loops you would be better off using `auto &` or `auto *` if you are using pointers so it is clear what is happening.

File: src/neta/connection.cpp [hunk 2]
```diff
@@ -152,10 +152,9 @@ int NETAConnectionNode::score(const SpeciesAtom* i, RefList<const SpeciesAtom>&
 
 	// Get directly connected atoms about 'i', excluding any that have already been matched
 	RefDataList<const SpeciesAtom, int> neighbours;
-	const PointerArray<SpeciesBond>& bonds = i->bonds();
-	for (int n=0; n<bonds.nItems(); ++n)
+	for (auto b : i->bonds())
 	{
-		const SpeciesAtom* j = bonds.at(n)->partner(i);
+		const SpeciesAtom* j = b->partner(i);
```
Comment (LamarMoore) [inline src/neta/connection.cpp]: `const auto *` modernize as you go along. Give a better name than `j` so you can have clarity.

File: src/neta/connection.cpp [hunk 3]
```diff
@@ -191,10 +190,10 @@ int NETAConnectionNode::score(const SpeciesAtom* i, RefList<const SpeciesAtom>&
 			// Now have a match, so break out of the loop
 			break;
 		}
-		if (atomScore == NETANode::NoMatch) for (int n=0; n<allowedAtomTypes_.nItems(); ++n)
+		if (atomScore == NETANode::NoMatch) for (auto t : allowedAtomTypes_)
 		{
 			// Evaluate the neighbour against the atom type
-			int typeScore = allowedAtomTypes_.at(n)->neta().score(j);
+			int typeScore = t->neta().score(j);
```
Comment (LamarMoore) [inline src/neta/connection.cpp]: auto

File: src/gui/viewer_render.cpp (hunk 1 above)
Comment (rprospero) [reply src/gui/viewer_render.cpp]: I've clarified where in the latest update the auto is working with pointers or references.

File: src/neta/connection.cpp (hunk 2 above)
Comment (rprospero) [reply src/neta/connection.cpp]: The latest update improves the names and attempts to modernize the code as it's changed.

**Comment by trisyoungs** [comment]
PointerArray can be completely removed rather than aliased to std::vector - pointless to retain the naming.

Something is badly broken, however, as all the system tests involving molecules with bonds now fail.

**Comment by rprospero** [comment]
I've isolated the commit where the test failure was introduced.  Just
tracking down the exact source of the bond issue.

**Comment by DanNixon** [comment]
As it is I think this is a good improvement.

As for what to do with the raw pointers, it is probably worth looking at moving interfaces to take references rather than pointers. Smart pointers would also work but I think you can avoid needing them based on what was described in the last meeting (all objects are created before any actual processing happens and are not before until exit).

**Comment by LamarMoore** [comment]
@rprospero is there an issue for addressing the use of pointer?

**Comment by rprospero** [comment]
I've created issue #244 to catalog the move to smarter pointers and #245 to track the move to STL containers.

**Pull request merged**